### PR TITLE
Initial support for ironic-standalone-operator

### PR DESCRIPTION
**Pull request by dtantsur**
Installs ironic-standalone-operator by default (but does not use it).

Adds a new option USE_IRSO (default false) to use the operator to provision Ironic.

Signed-off-by: Dmitry Tantsur <dtantsur@protonmail.com>

### CODE REVIEW

**Comment by dtantsur** [comment]
/test metal3-dev-env-integration-test-ubuntu-main

Should be solved now

**Review by tuminoid — COMMENTED** [review]
Its WIP, but two tiny nits, and one question.

The whole file needs cleanup on Bash so not picking on any of that, this PR is following convention just fine.

File: 03_launch_mgmt_cluster.sh [hunk 1]
```diff
@@ -562,6 +658,7 @@ if [ "${EPHEMERAL_CLUSTER}" != "tilt" ]; then
   launch_cluster_api_provider_metal3
   BMO_NAME_PREFIX="${NAMEPREFIX}"
   launch_baremetal_operator
+  launch_ironic_via_operator
```
[the remark this answers is not in the document]
Comment (dtantsur) [reply 03_launch_mgmt_cluster.sh]: Added in the latest revision

**Comment by tuminoid** [comment]
It looks like you've made good progress with the dev-env part and it is now actually failing mostly to IRSO implementation. Does that sound about right?

**Comment by dtantsur** [comment]
> It looks like you've made good progress with the dev-env part and it is now actually failing mostly to IRSO implementation. Does that sound about right?

Not exactly. The previous couple of failures are related to gluing BMO and newly deployed Ironic together. I presume the last failure was because of authentication, but without https://github.com/metal3-io/baremetal-operator/pull/2083 (now merged) I could not confirm that.

**Comment by dtantsur** [comment]
Okay, correcting: the very last failure may actually be a bug in IrSO.

File: lib/common.sh [hunk 2]
```diff
@@ -204,6 +206,8 @@ export BUILD_BMO_LOCALLY="${BUILD_BMO_LOCALLY:-false}"
 export BUILD_CAPI_LOCALLY="${BUILD_CAPI_LOCALLY:-false}"
 export BUILD_IRONIC_IMAGE_LOCALLY="${BUILD_IRONIC_IMAGE_LOCALLY:-false}"
 export BUILD_MARIADB_IMAGE_LOCALLY="${BUILD_MARIADB_IMAGE_LOCALLY:-false}"
+# FIXME(dtantsur): flip to false once everything merges
+export BUILD_IRSO_LOCALLY="${BUILD_IRSO_LOCALLY:-true}"
```
Comment (tuminoid) [inline lib/common.sh:210]: Just noting that dev-env best practice would be to set this to value it should be and then put required test config on top of `config_example.sh`.  This way it is clear what is prod code and what is test code.

Comment (dtantsur) [reply lib/common.sh:210]: Good to know, noted.

**Comment by dtantsur** [comment]
Yay, it works (fails only on the re-pivoting test).

File: 03_launch_mgmt_cluster.sh [hunk 3]
```diff
@@ -224,6 +224,99 @@ EOF
...
+RAMDISK_DOWNLOADER_IMAGE=$(get_component_image "${IPA_DOWNLOADER_LOCAL_IMAGE:-${IPA_DOWNLOADER_IMAGE}}")
+EOF
+
+  make -C "${IRSOPATH}" install deploy IMG="$(get_component_image "${IRSO_LOCAL_IMAGE:-${IRSO_IMAGE}}")"
+  kubectl wait --for=condition=Available --timeout=60s \
+    -n ironic-standalone-operator-system deployment/ironic-standalone-operator-controller-manager
+}
+
+launch_ironic_via_irso() {
+  if [ "${IRONIC_BASIC_AUTH}" != "true" ]; then
```
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:242]:
```suggestion
  if [[ "${IRONIC_BASIC_AUTH}" != "true" ]]; then
```

File: 03_launch_mgmt_cluster.sh [hunk 4]
```diff
@@ -311,56 +404,57 @@ function update_capm3_imports(){
   popd
 }
 
-#
-# Update the CAPM3 and BMO manifests to use local images as defined in variables
-#
-function update_component_image(){
-  IMPORT=$1
-  ORIG_IMAGE=$2
+function get_component_image(){
```
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:407]:
```suggestion
get_component_image() {
```

File: 03_launch_mgmt_cluster.sh [hunk 5]
```diff
@@ -311,56 +404,57 @@ function update_capm3_imports(){
...
     TMP_IMAGE_TAG="latest"
   fi
 
+  echo "${REGISTRY}/localimages/${TMP_IMAGE_NAME}:${TMP_IMAGE_TAG}"
+}
+
+#
+# Update the CAPM3 and BMO manifests to use local images as defined in variables
+#
+function update_component_image(){
```
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:424]:
```suggestion
update_component_image() {
```

File: config_example.sh [hunk 6]
```diff
@@ -213,3 +213,6 @@
 # To enable FakeIPA and run dev-env on a fake platform
 # export NODES_PLATFORM="fake"
 # export FAKE_IPA_IMAGE=192.168.111.1:5000/localimages/fake-ipa
+
+# Whether to use ironic-standalone-operator to deploy Ironic.
+# export USE_IRSO="true"
```
Comment (tuminoid) [inline config_example.sh:218]: I think this config is supposed to have default values shown, but quickly checking others, we have already mixed it. Up to you which one you want to have here.

**Review by tuminoid — COMMENTED** [review]
Completely optional nits. Looking solid, with documented variables and generally clean addition.

This will help greatly everyone to start experimenting with IRSO!

/approve

**Comment by dtantsur** [comment]
> Completely optional nits

I wonder if I should fix them in a follow-up or just wait for your reformatting of this file?

**Comment by tuminoid** [comment]
> > Completely optional nits
> 
> I wonder if I should fix them in a follow-up or just wait for your reformatting of this file?

If the tests pass, I can fix them later, or if you have any other need to push fixes to this PR, fix them as you go?

**Pull request merged**